execution/gethexec/sync_monitor.go

```diff
@@ -145,20 +98,47 @@ func (s *SyncMonitor) BlockMetadataByNumber(blockNum uint64) (common.BlockMetada
 	return nil, nil
 }
 
-func (s *SyncMonitor) StoreFinalityData(ctx context.Context, finalityData *arbutil.FinalityData) error {
-	s.finalityData.Store(finalityData)
+func (s *SyncMonitor) getFinalityBlock(
+	waitForBlockValidator bool,
+	validatedMsgCount *arbutil.MessageIndex,
+	finalityMsgCount arbutil.MessageIndex,
+) (*types.Block, error) {
+	if waitForBlockValidator {
+		if validatedMsgCount == nil {
+			return nil, errors.New("block validator not set")
+		}
+		if finalityMsgCount > *validatedMsgCount {
+			finalityMsgCount = *validatedMsgCount
+		}
+	}
+	finalityBlockNumber := s.exec.MessageIndexToBlockNumber(finalityMsgCount - 1)
+	finalityBlock := s.exec.bc.GetBlockByNumber(finalityBlockNumber)
+	if finalityBlock == nil {
+		return nil, errors.New("unable to get block by number")
+	}
+	return finalityBlock, nil
+}
 
-	finalizedBlockNumber, err := s.FinalizedBlockNumber(ctx)
-	if errors.Is(err, headerreader.ErrBlockNumberNotSupported) {
-		log.Warn("Finality not supported so not setting finalized block number")
-	} else if err == nil {
-		err = s.exec.SetFinalized(finalizedBlockNumber)
+func (s *SyncMonitor) SetFinalityData(ctx context.Context, finalityData *arbutil.FinalityData) error {
```

Comment:
separate PR:
I'm thinking about two contradicting options:
1) don't pass valid as part of finality data, only pass finalized & safe, let consensus side implement logic to limit finalized & safe by valid
2) do pass final, safe & valid, use valid to also replace "MarkValid" from the blockRecorder.
Not sure yet which makes more sense, specifically for scenario with multiple execution engines

Reply:
I decided to pass valid from consensus to execution because FinalizedBlockWaitForBlockValidator and SafeBlockWaitForBlockValidator configs are defined in the execution side.
So to implement option 1 we will need to do a breaking change in that regard, i.e., move these configs to the consensus side.

Reply:
makes sense. We may change it in the future, but it's a good deciding factor for now.